### PR TITLE
Update docstrings in `ParticleList`

### LINKED ISSUE #1712: Add examples for `ParticleList.sort` (e.g., sorting by charge number)
The docstring for `ParticleList.sort` doesn't contain any examples, which makes it harder to use.  It would be helpful to expand this docstring by adding an examples section.

A suggestion from @MKastek:
```Python
sorted(element_list, key=lambda x: atomic_number(x))
```

### PR DESCRIPTION
**Pull request by namurphy**
* Expanded the docstring for `ParticleList.sort`, in particular by adding sections for "Parameters", "Raises", "See Also", and "Examples".  This follows a suggestion from @MKastek.
* Change`ParticleList.sort` to use `super`, as suggested by @rocco8773.
* Update the docstrings of `ParticleList` and many of its methods/attributes.

Closes #1712.

### CODE REVIEW

File: plasmapy/particles/particle_collections.py [hunk 1]
```diff
@@ -22,35 +20,34 @@
 
 class ParticleList(collections.UserList):
     """
-    A `list` like collection of
-    `~plasmapy.particles.particle_class.Particle` and/or
-    `~plasmapy.particles.particle_class.CustomParticle` objects.
+    A `list` like collection of particle objects.
```
Comment (pheuer) [inline plasmapy/particles/particle_collections.py:23]: Why don't you want the links here?

File: plasmapy/particles/particle_collections.py [hunk 2]
```diff
@@ -217,16 +211,24 @@ def data(self) -> List[Union[Particle, CustomParticle]]:
         A `list` containing the particles contained in the
         |ParticleList| instance.
 
-        The `~plasmapy.particles.particle_collections.ParticleList.data`
-        attribute should not be modified directly.
+        .. important::
```
Comment (pheuer) [inline plasmapy/particles/particle_collections.py:214]: You already sort of enforce this by using a private attribute. You could do one better by adding a __setattr__ method for data that raises an exception?

File: plasmapy/particles/particle_collections.py [hunk 3]
```diff
@@ -292,24 +301,37 @@ def is_category(
     @property
     def charge_number(self) -> np.array:
         """
-        An array of the quantized charges of the particles, as
-        multiples of the elementary charge.
+        The charges of the particles in units of the elementary
+        positive charge.
```
Comment (pheuer) [inline plasmapy/particles/particle_collections.py:305]: Isn't the elementary charge e defined to be positive?

```suggestion
         charge.
```

File: plasmapy/particles/particle_collections.py [hunk 4]
```diff
@@ -321,16 +343,61 @@ def sort(self, key: Callable = None, reverse: bool = False):
...
+        To sort a |ParticleList| by atomic number, set ``key`` to
+        |atomic_number|.
+
+        >>> from plasmapy.particles import ParticleList, atomic_number
+        >>> elements = ParticleList(["Li", "H", "He"])
+        >>> elements.sort(key=atomic_number)
+        >>> print(elements)
+        ParticleList(['H', 'He', 'Li'])
+
+        We can also create a function that we then pass to ``key``. In
```
Comment (pheuer) [inline plasmapy/particles/particle_collections.py:377]: I don't quite understand this example yet - how do the three quantities coming from this function define the sorting? Does it sort based off of the first one and then break ties with the next one? I think that could be said a tad more explicitly.

File: plasmapy/particles/particle_collections.py [hunk 5]
```diff
@@ -321,16 +343,61 @@ def sort(self, key: Callable = None, reverse: bool = False):
         """
         Sort the |ParticleList| in-place.
 
-        For more information, refer to the documentation for `list.sort`.
+        Parameters
+        ----------
+        key : callable
+            A function of one argument that is used to extract a
```
Comment (pheuer) [inline plasmapy/particles/particle_collections.py:349]:
```suggestion
            A function that accepts one |Particle| argument that is used to extract a
```

File: plasmapy/particles/particle_collections.py (hunk 1 above)
Comment (namurphy) [reply plasmapy/particles/particle_collections.py:23]: Good question. In retrospect, it does seem a better idea to link to `Particle` and `CustomParticle`, doesn't it?

File: plasmapy/particles/particle_collections.py (hunk 2 above)
Comment (namurphy) [reply plasmapy/particles/particle_collections.py:214]: This one is kind of complicated.  `ParticleList` inherits `data` from `collections.UserList`, so we would need to keep `data` itself public to be consistent with the [Liskov Substitution Principle](https://reflectoring.io/lsp-explained/).  Even if we make `data` a `property`, it is still a `list` so we can do operations like:
```Python
>>> particle_list = ParticleList([])
>>> particle_list.data.append("not a particle")
```
The admonition is to warn against operations like that.

Comment (pheuer) [reply plasmapy/particles/particle_collections.py:214]: Ok, got it. 

I think I've run into this problem before - would be nice if there were an abstract datatype like a list that isn't modifiable except by subclass methods. Like a 'secret list'

**Review by pheuer — APPROVED** [review]
Great, otherwise looks like straightforward and good improvements!

**Comment by namurphy** [comment]
Thank you for the review, @pheuer!

**Pull request merged**